fatal lld errors don't get printed

If you run into a fatal error in LLD (i.e. if a section's sh_size is not a multiple of its entry size), then lld will print the error and then call `_exit`.  The fact that `lld` exits is a problem on its own, but this issue is for the missing error message.

With strace, I noticed that around the time that the fatal error message should be printed, LLD calls lseek on `stdout` and gets a `-ESPIPE` error.  This may or may not have something to do with why the error message does not get printed.

I was hitting this while making some changes, I had an error that caused the compiler to generate an object file with an invalid entry for the section size.  You can reproduce this case by applying this patch:

```patch
--- deps/lld/ELF/InputFiles.cpp	2019-12-01 13:00:48.384630138 -0700
+++ deps/lld/ELF/InputFiles.cpp	2019-12-11 17:06:39.466770163 -0700
@@ -515,7 +515,7 @@
   uint64_t entSize = sec.sh_entsize;
   if (entSize == 0)
     return false;
-  if (sec.sh_size % entSize)
+  if (true) // always fail
     fatal(toString(this) +
           ": SHF_MERGE section size must be a multiple of sh_entsize");
```

After applying this patch, run `make install` and you'll see that it fails but you won't get any helpful error message.  You can use `make install VERBOSE=1" to get more output.  Here's the full output of me running `make install VERBOSE=1`:

```
[marler8997@nixos:~/git/zig/build]$ make install VERBOSE=1
/nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -S/home/marler8997/git/zig -B/home/marler8997/git/zig/build --check-build-system CMakeFiles/Makefile.cmake 0
/nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_progress_start /home/marler8997/git/zig/build/CMakeFiles /home/marler8997/git/zig/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_mingw.dir/build.make CMakeFiles/embedded_lld_mingw.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_lld_mingw.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_mingw.dir/build.make CMakeFiles/embedded_lld_mingw.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_lld_mingw.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[  1%] Built target embedded_lld_mingw
make -f CMakeFiles/embedded_softfloat.dir/build.make CMakeFiles/embedded_softfloat.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_softfloat.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_softfloat.dir/build.make CMakeFiles/embedded_softfloat.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_softfloat.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 47%] Built target embedded_softfloat
make -f CMakeFiles/embedded_lld_lib.dir/build.make CMakeFiles/embedded_lld_lib.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_lld_lib.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_lib.dir/build.make CMakeFiles/embedded_lld_lib.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_lld_lib.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 62%] Built target embedded_lld_lib
make -f CMakeFiles/zig_cpp.dir/build.make CMakeFiles/zig_cpp.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/zig_cpp.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/zig_cpp.dir/build.make CMakeFiles/zig_cpp.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/zig_cpp.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 65%] Built target zig_cpp
make -f CMakeFiles/opt_c_util.dir/build.make CMakeFiles/opt_c_util.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/opt_c_util.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/opt_c_util.dir/build.make CMakeFiles/opt_c_util.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/opt_c_util.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 66%] Built target opt_c_util
make -f CMakeFiles/embedded_lld_wasm.dir/build.make CMakeFiles/embedded_lld_wasm.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_lld_wasm.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_wasm.dir/build.make CMakeFiles/embedded_lld_wasm.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_lld_wasm.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 70%] Built target embedded_lld_wasm
make -f CMakeFiles/embedded_lld_elf.dir/build.make CMakeFiles/embedded_lld_elf.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_lld_elf.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_elf.dir/build.make CMakeFiles/embedded_lld_elf.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_lld_elf.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 84%] Built target embedded_lld_elf
make -f CMakeFiles/embedded_lld_coff.dir/build.make CMakeFiles/embedded_lld_coff.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/embedded_lld_coff.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/embedded_lld_coff.dir/build.make CMakeFiles/embedded_lld_coff.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/embedded_lld_coff.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 89%] Built target embedded_lld_coff
make -f CMakeFiles/compiler.dir/build.make CMakeFiles/compiler.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/compiler.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/compiler.dir/build.make CMakeFiles/compiler.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/compiler.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 98%] Built target compiler
make -f CMakeFiles/zig0.dir/build.make CMakeFiles/zig0.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/zig0.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/zig0.dir/build.make CMakeFiles/zig0.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
make[2]: Nothing to be done for 'CMakeFiles/zig0.dir/build'.
make[2]: Leaving directory '/home/marler8997/git/zig/build'
[ 99%] Built target zig0
make -f CMakeFiles/zig_build_libuserland.dir/build.make CMakeFiles/zig_build_libuserland.dir/depend
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig/build && /nix/store/hl5d3b9f9psdh54z42ychqd23y8iwiwv-cmake-3.13.4/bin/cmake -E cmake_depends "Unix Makefiles" /home/marler8997/git/zig /home/marler8997/git/zig /home/marler8997/git/zig/build /home/marler8997/git/zig/build /home/marler8997/git/zig/build/CMakeFiles/zig_build_libuserland.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make -f CMakeFiles/zig_build_libuserland.dir/build.make CMakeFiles/zig_build_libuserland.dir/build
make[2]: Entering directory '/home/marler8997/git/zig/build'
cd /home/marler8997/git/zig && /home/marler8997/git/zig/build/zig0 build --override-lib-dir /home/marler8997/git/zig/lib libuserland install -Doutput-dir=/home/marler8997/git/zig/build -Drelease=false -Dlib-files-only --prefix /home/marler8997/git/zig/build
make[2]: *** [CMakeFiles/zig_build_libuserland.dir/build.make:57: CMakeFiles/zig_build_libuserland] Error 1
make[2]: Leaving directory '/home/marler8997/git/zig/build'
make[1]: *** [CMakeFiles/Makefile2:265: CMakeFiles/zig_build_libuserland.dir/all] Error 2
make[1]: Leaving directory '/home/marler8997/git/zig/build'
make: *** [Makefile:130: all] Error 2

```

Because of the patch, we should be seeing an error message like this
```
build.o: SHF_MERGE section size must be a multiple of sh_entsize
```

but this error message is nowhere to be found.

Note that I can fix the issue with this patch:
```patch
--- deps/lld/Common/ErrorHandler.cpp
+++ deps/lld/Common/ErrorHandler.cpp
@@ -172,7 +172,13 @@
   ++errorCount;
 }
 
+#include <iostream>
 void ErrorHandler::fatal(const Twine &msg) {
   error(msg);
+  std::string msgStr = msg.str();
+  const char *msgCStr = msgStr.c_str();
+  fprintf(stderr, "FATAL ERROR: %s\n", msgCStr);
+  fflush(stderr);
+
   exitLld(1);
 }

```

With this I see get following output when I run `make install`:
```
[marler8997@nixos:~/git/zig/build]$ make install
[  1%] Built target embedded_lld_mingw
[ 47%] Built target embedded_softfloat
Scanning dependencies of target embedded_lld_lib
[ 48%] Building CXX object CMakeFiles/embedded_lld_lib.dir/deps/lld/Common/ErrorHandler.cpp.o
[ 49%] Linking CXX static library zig_cpp/libembedded_lld_lib.a
[ 62%] Built target embedded_lld_lib
[ 65%] Built target zig_cpp
[ 66%] Built target opt_c_util
[ 70%] Built target embedded_lld_wasm
[ 84%] Built target embedded_lld_elf
[ 89%] Built target embedded_lld_coff
[ 98%] Built target compiler
[ 98%] Linking CXX executable zig0
[ 99%] Built target zig0
FATAL ERROR: /home/marler8997/git/zig/zig-cache/o/eH0KG2eA5RsCDDdKpa4Y7kNmCVIazbMUOZJHKtIsQqGOqkZ_wuHOfasJbeNtf1Lw/build.o: SHF_MERGE section size must be a multiple of sh_entsize
make[2]: *** [CMakeFiles/zig_build_libuserland.dir/build.make:57: CMakeFiles/zig_build_libuserland] Error 1
make[1]: *** [CMakeFiles/Makefile2:265: CMakeFiles/zig_build_libuserland.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

So the problem is that the `error(msg)` call is not getting printed.  A separate issue is that LLD is exiting, I don't think it should be doing that, but it looks like that's what the code always does, there's no condition that checks if it's running in library mode before it calls the "_exit" system call.

--- a/lib/std/testing.zig
+++ b/lib/std/testing.zig
@@ -247,6 +247,7 @@ test "expectWithinEpsilon" {
 /// This function is intended to be used only in tests. When the two slices are not
 /// equal, prints diagnostics to stderr to show exactly how they are not equal,
 /// then aborts.
+/// If your inputs are UTF-8 encoded strings, consider calling `expectEqualStrings` instead.
 pub fn expectEqualSlices(comptime T: type, expected: []const T, actual: []const T) void {
     // TODO better printing of the difference
     // If the arrays are small enough we could print the whole thing
@@ -366,6 +367,26 @@ pub fn expectEqualStrings(expected: []const u8, actual: []const u8) void {
 
         @panic("test failure");
     }
+}
+
+pub fn expectStringEndsWith(actual: []const u8, expected_ends_with: []const u8) void {
+    if (std.mem.endsWith(u8, actual, expected_ends_with))
+        return;
+
+    const shortened_actual = if (actual.len >= expected_ends_with.len)
+        actual[0..expected_ends_with.len]
+    else
+        actual;
+
+    print("\n====== expected to end with: =========\n", .{});
+    printWithVisibleNewlines(expected_ends_with);
+    print("\n====== instead ended with: ===========\n", .{});
+    printWithVisibleNewlines(shortened_actual);
+    print("\n========= full output: ==============\n", .{});
+    printWithVisibleNewlines(actual);
+    print("\n======================================\n", .{});
+
+    @panic("test failure");
 }
 
 fn printIndicatorLine(source: []const u8, indicator_index: usize) void {

--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -1804,7 +1804,7 @@ fn updateCObject(comp: *Compilation, c_object: *CObject, c_comp_progress_node: *
                     if (comp.clang_preprocessor_mode == .stdout)
                         std.process.exit(0);
                 },
-                else => std.process.exit(1),
+                else => std.process.abort(),
             }
         } else {
             child.stdin_behavior = .Ignore;

--- a/src/link/Coff.zig
+++ b/src/link/Coff.zig
@@ -907,11 +907,10 @@ fn linkWithLLD(self: *Coff, comp: *Compilation) !void {
         // Create an LLD command line and invoke it.
         var argv = std.ArrayList([]const u8).init(self.base.allocator);
         defer argv.deinit();
-        // The first argument is ignored as LLD is called as a library, set it
-        // anyway to the correct LLD driver name for this target so that it's
-        // correctly printed when `verbose_link` is true. This is needed for some
-        // tools such as CMake when Zig is used as C compiler.
-        try argv.append("lld-link");
+        // We will invoke ourselves as a child process to gain access to LLD.
+        // This is necessary because LLD does not behave properly as a library -
+        // it calls exit() and does not reset all global data between invocations.
+        try argv.appendSlice(&[_][]const u8{ comp.self_exe_path.?, "lld-link" });
 
         try argv.append("-ERRORLIMIT:0");
         try argv.append("-NOLOGO");
@@ -1149,45 +1148,51 @@ fn linkWithLLD(self: *Coff, comp: *Compilation) !void {
         }
 
         if (self.base.options.verbose_link) {
-            Compilation.dump_argv(argv.items);
+            // Skip over our own name so that the LLD linker name is the first argv item.
+            Compilation.dump_argv(argv.items[1..]);
         }
 
-        const new_argv = try arena.allocSentinel(?[*:0]const u8, argv.items.len, null);
-        for (argv.items) |arg, i| {
-            new_argv[i] = try arena.dupeZ(u8, arg);
+        // Sadly, we must run LLD as a child process because it does not behave
+        // properly as a library. One exception is if we are running in passthrough
+        // mode, which means Clang / LLD should inherit stdio and are allowed to
+        // crash zig directly.
+        if (comp.clang_passthrough_mode) {
+            return @import("../main.zig").punt_to_lld(arena, argv.items);
         }
 
-        var stderr_context: LLDContext = .{
-            .coff = self,
-            .data = std.ArrayList(u8).init(self.base.allocator),
+        const child = try std.ChildProcess.init(argv.items, arena);
+        defer child.deinit();
+
+        child.stdin_behavior = .Ignore;
+        child.stdout_behavior = .Ignore;
+        child.stderr_behavior = .Pipe;
+
+        try child.spawn();
+
+        const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+
+        const term = child.wait() catch |err| {
+            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+            return error.UnableToSpawnSelf;
         };
-        defer stderr_context.data.deinit();
-        var stdout_context: LLDContext = .{
-            .coff = self,
-            .data = std.ArrayList(u8).init(self.base.allocator),
-        };
-        defer stdout_context.data.deinit();
-        const llvm = @import("../llvm.zig");
-        const ok = llvm.Link(
-            .COFF,
-            new_argv.ptr,
-            new_argv.len,
-            append_diagnostic,
-            @ptrToInt(&stdout_context),
-            @ptrToInt(&stderr_context),
-        );
-        if (stderr_context.oom or stdout_context.oom) return error.OutOfMemory;
-        if (stdout_context.data.items.len != 0) {
-            std.log.warn("unexpected LLD stdout: {}", .{stdout_context.data.items});
+
+        switch (term) {
+            .Exited => |code| {
+                if (code != 0) {
+                    // TODO parse this output and surface with the Compilation API rather than
+                    // directly outputting to stderr here.
+                    std.debug.print("{s}", .{stderr});
+                    return error.LLDReportedFailure;
+                }
+            },
+            else => {
+                log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                return error.LLDCrashed;
+            },
         }
-        if (!ok) {
-            // TODO parse this output and surface with the Compilation API rather than
-            // directly outputting to stderr here.
-            std.debug.print("{}", .{stderr_context.data.items});
-            return error.LLDReportedFailure;
-        }
-        if (stderr_context.data.items.len != 0) {
-            std.log.warn("unexpected LLD stderr: {}", .{stderr_context.data.items});
+
+        if (stderr.len != 0) {
+            std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
         }
     }
 
@@ -1205,20 +1210,6 @@ fn linkWithLLD(self: *Coff, comp: *Compilation) !void {
         // other processes clobbering it.
         self.base.lock = man.toOwnedLock();
     }
-}
-
-const LLDContext = struct {
-    data: std.ArrayList(u8),
-    coff: *Coff,
-    oom: bool = false,
-};
-
-fn append_diagnostic(context: usize, ptr: [*]const u8, len: usize) callconv(.C) void {
-    const lld_context = @intToPtr(*LLDContext, context);
-    const msg = ptr[0..len];
-    lld_context.data.appendSlice(msg) catch |err| switch (err) {
-        error.OutOfMemory => lld_context.oom = true,
-    };
 }
 
 pub fn getDeclVAddr(self: *Coff, decl: *const Module.Decl) u64 {

--- a/src/link/Coff.zig
+++ b/src/link/Coff.zig
@@ -1153,46 +1153,60 @@ fn linkWithLLD(self: *Coff, comp: *Compilation) !void {
         }
 
         // Sadly, we must run LLD as a child process because it does not behave
-        // properly as a library. One exception is if we are running in passthrough
-        // mode, which means Clang / LLD should inherit stdio and are allowed to
-        // crash zig directly.
-        if (comp.clang_passthrough_mode) {
-            return @import("../main.zig").punt_to_lld(arena, argv.items);
-        }
-
+        // properly as a library.
         const child = try std.ChildProcess.init(argv.items, arena);
         defer child.deinit();
 
-        child.stdin_behavior = .Ignore;
-        child.stdout_behavior = .Ignore;
-        child.stderr_behavior = .Pipe;
+        if (comp.clang_passthrough_mode) {
+            child.stdin_behavior = .Inherit;
+            child.stdout_behavior = .Inherit;
+            child.stderr_behavior = .Inherit;
 
-        try child.spawn();
+            const term = child.spawnAndWait() catch |err| {
+                log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+                return error.UnableToSpawnSelf;
+            };
+            switch (term) {
+                .Exited => |code| {
+                    if (code != 0) {
+                        // TODO https://github.com/ziglang/zig/issues/6342
+                        std.process.exit(1);
+                    }
+                },
+                else => std.process.abort(),
+            }
+        } else {
+            child.stdin_behavior = .Ignore;
+            child.stdout_behavior = .Ignore;
+            child.stderr_behavior = .Pipe;
 
-        const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+            try child.spawn();
 
-        const term = child.wait() catch |err| {
-            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
-            return error.UnableToSpawnSelf;
-        };
+            const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
 
-        switch (term) {
-            .Exited => |code| {
-                if (code != 0) {
-                    // TODO parse this output and surface with the Compilation API rather than
-                    // directly outputting to stderr here.
-                    std.debug.print("{s}", .{stderr});
-                    return error.LLDReportedFailure;
-                }
-            },
-            else => {
-                log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
-                return error.LLDCrashed;
-            },
-        }
+            const term = child.wait() catch |err| {
+                log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+                return error.UnableToSpawnSelf;
+            };
 
-        if (stderr.len != 0) {
-            std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+            switch (term) {
+                .Exited => |code| {
+                    if (code != 0) {
+                        // TODO parse this output and surface with the Compilation API rather than
+                        // directly outputting to stderr here.
+                        std.debug.print("{s}", .{stderr});
+                        return error.LLDReportedFailure;
+                    }
+                },
+                else => {
+                    log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                    return error.LLDCrashed;
+                },
+            }
+
+            if (stderr.len != 0) {
+                std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+            }
         }
     }
 

--- a/src/link/Elf.zig
+++ b/src/link/Elf.zig
@@ -1632,46 +1632,60 @@ fn linkWithLLD(self: *Elf, comp: *Compilation) !void {
     }
 
     // Sadly, we must run LLD as a child process because it does not behave
-    // properly as a library. One exception is if we are running in passthrough
-    // mode, which means Clang / LLD should inherit stdio and are allowed to
-    // crash zig directly.
-    if (comp.clang_passthrough_mode) {
-        return @import("../main.zig").punt_to_lld(arena, argv.items);
-    }
-
+    // properly as a library.
     const child = try std.ChildProcess.init(argv.items, arena);
     defer child.deinit();
 
-    child.stdin_behavior = .Ignore;
-    child.stdout_behavior = .Ignore;
-    child.stderr_behavior = .Pipe;
+    if (comp.clang_passthrough_mode) {
+        child.stdin_behavior = .Inherit;
+        child.stdout_behavior = .Inherit;
+        child.stderr_behavior = .Inherit;
 
-    try child.spawn();
+        const term = child.spawnAndWait() catch |err| {
+            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+            return error.UnableToSpawnSelf;
+        };
+        switch (term) {
+            .Exited => |code| {
+                if (code != 0) {
+                    // TODO https://github.com/ziglang/zig/issues/6342
+                    std.process.exit(1);
+                }
+            },
+            else => std.process.abort(),
+        }
+    } else {
+        child.stdin_behavior = .Ignore;
+        child.stdout_behavior = .Ignore;
+        child.stderr_behavior = .Pipe;
 
-    const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+        try child.spawn();
 
-    const term = child.wait() catch |err| {
-        log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
-        return error.UnableToSpawnSelf;
-    };
+        const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
 
-    switch (term) {
-        .Exited => |code| {
-            if (code != 0) {
-                // TODO parse this output and surface with the Compilation API rather than
-                // directly outputting to stderr here.
-                std.debug.print("{s}", .{stderr});
-                return error.LLDReportedFailure;
-            }
-        },
-        else => {
-            log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
-            return error.LLDCrashed;
-        },
-    }
+        const term = child.wait() catch |err| {
+            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+            return error.UnableToSpawnSelf;
+        };
 
-    if (stderr.len != 0) {
-        std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+        switch (term) {
+            .Exited => |code| {
+                if (code != 0) {
+                    // TODO parse this output and surface with the Compilation API rather than
+                    // directly outputting to stderr here.
+                    std.debug.print("{s}", .{stderr});
+                    return error.LLDReportedFailure;
+                }
+            },
+            else => {
+                log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                return error.LLDCrashed;
+            },
+        }
+
+        if (stderr.len != 0) {
+            std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+        }
     }
 
     if (!self.base.options.disable_lld_caching) {

--- a/src/link/Elf.zig
+++ b/src/link/Elf.zig
@@ -1360,11 +1360,10 @@ fn linkWithLLD(self: *Elf, comp: *Compilation) !void {
     // Create an LLD command line and invoke it.
     var argv = std.ArrayList([]const u8).init(self.base.allocator);
     defer argv.deinit();
-    // The first argument is ignored as LLD is called as a library, set it
-    // anyway to the correct LLD driver name for this target so that it's
-    // correctly printed when `verbose_link` is true. This is needed for some
-    // tools such as CMake when Zig is used as C compiler.
-    try argv.append("ld.lld");
+    // We will invoke ourselves as a child process to gain access to LLD.
+    // This is necessary because LLD does not behave properly as a library -
+    // it calls exit() and does not reset all global data between invocations.
+    try argv.appendSlice(&[_][]const u8{ comp.self_exe_path.?, "ld.lld" });
     if (is_obj) {
         try argv.append("-r");
     }
@@ -1628,46 +1627,51 @@ fn linkWithLLD(self: *Elf, comp: *Compilation) !void {
     }
 
     if (self.base.options.verbose_link) {
-        Compilation.dump_argv(argv.items);
+        // Skip over our own name so that the LLD linker name is the first argv item.
+        Compilation.dump_argv(argv.items[1..]);
     }
 
-    // Oh, snapplesauce! We need null terminated argv.
-    const new_argv = try arena.allocSentinel(?[*:0]const u8, argv.items.len, null);
-    for (argv.items) |arg, i| {
-        new_argv[i] = try arena.dupeZ(u8, arg);
+    // Sadly, we must run LLD as a child process because it does not behave
+    // properly as a library. One exception is if we are running in passthrough
+    // mode, which means Clang / LLD should inherit stdio and are allowed to
+    // crash zig directly.
+    if (comp.clang_passthrough_mode) {
+        return @import("../main.zig").punt_to_lld(arena, argv.items);
     }
 
-    var stderr_context: LLDContext = .{
-        .elf = self,
-        .data = std.ArrayList(u8).init(self.base.allocator),
+    const child = try std.ChildProcess.init(argv.items, arena);
+    defer child.deinit();
+
+    child.stdin_behavior = .Ignore;
+    child.stdout_behavior = .Ignore;
+    child.stderr_behavior = .Pipe;
+
+    try child.spawn();
+
+    const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+
+    const term = child.wait() catch |err| {
+        log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+        return error.UnableToSpawnSelf;
     };
-    defer stderr_context.data.deinit();
-    var stdout_context: LLDContext = .{
-        .elf = self,
-        .data = std.ArrayList(u8).init(self.base.allocator),
-    };
-    defer stdout_context.data.deinit();
-    const llvm = @import("../llvm.zig");
-    const ok = llvm.Link(
-        .ELF,
-        new_argv.ptr,
-        new_argv.len,
-        append_diagnostic,
-        @ptrToInt(&stdout_context),
-        @ptrToInt(&stderr_context),
-    );
-    if (stderr_context.oom or stdout_context.oom) return error.OutOfMemory;
-    if (stdout_context.data.items.len != 0) {
-        std.log.warn("unexpected LLD stdout: {}", .{stdout_context.data.items});
+
+    switch (term) {
+        .Exited => |code| {
+            if (code != 0) {
+                // TODO parse this output and surface with the Compilation API rather than
+                // directly outputting to stderr here.
+                std.debug.print("{s}", .{stderr});
+                return error.LLDReportedFailure;
+            }
+        },
+        else => {
+            log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+            return error.LLDCrashed;
+        },
     }
-    if (!ok) {
-        // TODO parse this output and surface with the Compilation API rather than
-        // directly outputting to stderr here.
-        std.debug.print("{}", .{stderr_context.data.items});
-        return error.LLDReportedFailure;
-    }
-    if (stderr_context.data.items.len != 0) {
-        std.log.warn("unexpected LLD stderr: {}", .{stderr_context.data.items});
+
+    if (stderr.len != 0) {
+        std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
     }
 
     if (!self.base.options.disable_lld_caching) {
@@ -1684,20 +1688,6 @@ fn linkWithLLD(self: *Elf, comp: *Compilation) !void {
         // other processes clobbering it.
         self.base.lock = man.toOwnedLock();
     }
-}
-
-const LLDContext = struct {
-    data: std.ArrayList(u8),
-    elf: *Elf,
-    oom: bool = false,
-};
-
-fn append_diagnostic(context: usize, ptr: [*]const u8, len: usize) callconv(.C) void {
-    const lld_context = @intToPtr(*LLDContext, context);
-    const msg = ptr[0..len];
-    lld_context.data.appendSlice(msg) catch |err| switch (err) {
-        error.OutOfMemory => lld_context.oom = true,
-    };
 }
 
 fn writeDwarfAddrAssumeCapacity(self: *Elf, buf: *std.ArrayList(u8), addr: u64) void {

--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -489,12 +489,10 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
         if (self.base.options.system_linker_hack) {
             try argv.append("ld");
         } else {
-            // The first argument is ignored as LLD is called as a library, set
-            // it anyway to the correct LLD driver name for this target so that
-            // it's correctly printed when `verbose_link` is true. This is
-            // needed for some tools such as CMake when Zig is used as C
-            // compiler.
-            try argv.append("ld64");
+            // We will invoke ourselves as a child process to gain access to LLD.
+            // This is necessary because LLD does not behave properly as a library -
+            // it calls exit() and does not reset all global data between invocations.
+            try argv.appendSlice(&[_][]const u8{ comp.self_exe_path.?, "ld64.lld" });
 
             try argv.append("-error-limit");
             try argv.append("0");
@@ -660,7 +658,9 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
         }
 
         if (self.base.options.verbose_link) {
-            Compilation.dump_argv(argv.items);
+            // Potentially skip over our own name so that the LLD linker name is the first argv item.
+            const adjusted_argv = if (self.base.options.system_linker_hack) argv.items else argv.items[1..];
+            Compilation.dump_argv(adjusted_argv);
         }
 
         // TODO https://github.com/ziglang/zig/issues/6971
@@ -685,42 +685,47 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
                 return error.LDReportedFailure;
             }
         } else {
-            const new_argv = try arena.allocSentinel(?[*:0]const u8, argv.items.len, null);
-            for (argv.items) |arg, i| {
-                new_argv[i] = try arena.dupeZ(u8, arg);
+            // Sadly, we must run LLD as a child process because it does not behave
+            // properly as a library. One exception is if we are running in passthrough
+            // mode, which means Clang / LLD should inherit stdio and are allowed to
+            // crash zig directly.
+            if (comp.clang_passthrough_mode) {
+                return @import("../main.zig").punt_to_lld(arena, argv.items);
             }
 
-            var stderr_context: LLDContext = .{
-                .macho = self,
-                .data = std.ArrayList(u8).init(self.base.allocator),
+            const child = try std.ChildProcess.init(argv.items, arena);
+            defer child.deinit();
+
+            child.stdin_behavior = .Ignore;
+            child.stdout_behavior = .Ignore;
+            child.stderr_behavior = .Pipe;
+
+            try child.spawn();
+
+            const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+
+            const term = child.wait() catch |err| {
+                log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+                return error.UnableToSpawnSelf;
             };
-            defer stderr_context.data.deinit();
-            var stdout_context: LLDContext = .{
-                .macho = self,
-                .data = std.ArrayList(u8).init(self.base.allocator),
-            };
-            defer stdout_context.data.deinit();
-            const llvm = @import("../llvm.zig");
-            const ok = llvm.Link(
-                .MachO,
-                new_argv.ptr,
-                new_argv.len,
-                append_diagnostic,
-                @ptrToInt(&stdout_context),
-                @ptrToInt(&stderr_context),
-            );
-            if (stderr_context.oom or stdout_context.oom) return error.OutOfMemory;
-            if (stdout_context.data.items.len != 0) {
-                std.log.warn("unexpected LLD stdout: {}", .{stdout_context.data.items});
+
+            switch (term) {
+                .Exited => |code| {
+                    if (code != 0) {
+                        // TODO parse this output and surface with the Compilation API rather than
+                        // directly outputting to stderr here.
+                        std.debug.print("{s}", .{stderr});
+                        return error.LLDReportedFailure;
+                    }
+                },
+                else => {
+                    log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                    return error.LLDCrashed;
+                },
             }
-            if (!ok) {
-                // TODO parse this output and surface with the Compilation API rather than
-                // directly outputting to stderr here.
-                std.log.err("{}", .{stderr_context.data.items});
-                return error.LLDReportedFailure;
-            }
-            if (stderr_context.data.items.len != 0) {
-                std.log.warn("unexpected LLD stderr: {}", .{stderr_context.data.items});
+
+            if (stderr.len != 0) {
+                std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
             }
 
             // At this stage, LLD has done its job. It is time to patch the resultant
@@ -783,20 +788,6 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
         // other processes clobbering it.
         self.base.lock = man.toOwnedLock();
     }
-}
-
-const LLDContext = struct {
-    data: std.ArrayList(u8),
-    macho: *MachO,
-    oom: bool = false,
-};
-
-fn append_diagnostic(context: usize, ptr: [*]const u8, len: usize) callconv(.C) void {
-    const lld_context = @intToPtr(*LLDContext, context);
-    const msg = ptr[0..len];
-    lld_context.data.appendSlice(msg) catch |err| switch (err) {
-        error.OutOfMemory => lld_context.oom = true,
-    };
 }
 
 fn darwinArchString(arch: std.Target.Cpu.Arch) []const u8 {

--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -686,46 +686,60 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
             }
         } else {
             // Sadly, we must run LLD as a child process because it does not behave
-            // properly as a library. One exception is if we are running in passthrough
-            // mode, which means Clang / LLD should inherit stdio and are allowed to
-            // crash zig directly.
-            if (comp.clang_passthrough_mode) {
-                return @import("../main.zig").punt_to_lld(arena, argv.items);
-            }
-
+            // properly as a library.
             const child = try std.ChildProcess.init(argv.items, arena);
             defer child.deinit();
 
-            child.stdin_behavior = .Ignore;
-            child.stdout_behavior = .Ignore;
-            child.stderr_behavior = .Pipe;
+            if (comp.clang_passthrough_mode) {
+                child.stdin_behavior = .Inherit;
+                child.stdout_behavior = .Inherit;
+                child.stderr_behavior = .Inherit;
 
-            try child.spawn();
+                const term = child.spawnAndWait() catch |err| {
+                    log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+                    return error.UnableToSpawnSelf;
+                };
+                switch (term) {
+                    .Exited => |code| {
+                        if (code != 0) {
+                            // TODO https://github.com/ziglang/zig/issues/6342
+                            std.process.exit(1);
+                        }
+                    },
+                    else => std.process.abort(),
+                }
+            } else {
+                child.stdin_behavior = .Ignore;
+                child.stdout_behavior = .Ignore;
+                child.stderr_behavior = .Pipe;
 
-            const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+                try child.spawn();
 
-            const term = child.wait() catch |err| {
-                log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
-                return error.UnableToSpawnSelf;
-            };
+                const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
 
-            switch (term) {
-                .Exited => |code| {
-                    if (code != 0) {
-                        // TODO parse this output and surface with the Compilation API rather than
-                        // directly outputting to stderr here.
-                        std.debug.print("{s}", .{stderr});
-                        return error.LLDReportedFailure;
-                    }
-                },
-                else => {
-                    log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
-                    return error.LLDCrashed;
-                },
-            }
+                const term = child.wait() catch |err| {
+                    log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+                    return error.UnableToSpawnSelf;
+                };
 
-            if (stderr.len != 0) {
-                std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+                switch (term) {
+                    .Exited => |code| {
+                        if (code != 0) {
+                            // TODO parse this output and surface with the Compilation API rather than
+                            // directly outputting to stderr here.
+                            std.debug.print("{s}", .{stderr});
+                            return error.LLDReportedFailure;
+                        }
+                    },
+                    else => {
+                        log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                        return error.LLDCrashed;
+                    },
+                }
+
+                if (stderr.len != 0) {
+                    std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+                }
             }
 
             // At this stage, LLD has done its job. It is time to patch the resultant

--- a/src/link/Wasm.zig
+++ b/src/link/Wasm.zig
@@ -345,11 +345,10 @@ fn linkWithLLD(self: *Wasm, comp: *Compilation) !void {
     // Create an LLD command line and invoke it.
     var argv = std.ArrayList([]const u8).init(self.base.allocator);
     defer argv.deinit();
-    // The first argument is ignored as LLD is called as a library, set it
-    // anyway to the correct LLD driver name for this target so that it's
-    // correctly printed when `verbose_link` is true. This is needed for some
-    // tools such as CMake when Zig is used as C compiler.
-    try argv.append("ld-wasm");
+    // We will invoke ourselves as a child process to gain access to LLD.
+    // This is necessary because LLD does not behave properly as a library -
+    // it calls exit() and does not reset all global data between invocations.
+    try argv.appendSlice(&[_][]const u8{ comp.self_exe_path.?, "wasm-ld" });
     if (is_obj) {
         try argv.append("-r");
     }
@@ -399,45 +398,51 @@ fn linkWithLLD(self: *Wasm, comp: *Compilation) !void {
     }
 
     if (self.base.options.verbose_link) {
-        Compilation.dump_argv(argv.items);
+        // Skip over our own name so that the LLD linker name is the first argv item.
+        Compilation.dump_argv(argv.items[1..]);
     }
 
-    const new_argv = try arena.allocSentinel(?[*:0]const u8, argv.items.len, null);
-    for (argv.items) |arg, i| {
-        new_argv[i] = try arena.dupeZ(u8, arg);
+    // Sadly, we must run LLD as a child process because it does not behave
+    // properly as a library. One exception is if we are running in passthrough
+    // mode, which means Clang / LLD should inherit stdio and are allowed to
+    // crash zig directly.
+    if (comp.clang_passthrough_mode) {
+        return @import("../main.zig").punt_to_lld(arena, argv.items);
     }
 
-    var stderr_context: LLDContext = .{
-        .wasm = self,
-        .data = std.ArrayList(u8).init(self.base.allocator),
+    const child = try std.ChildProcess.init(argv.items, arena);
+    defer child.deinit();
+
+    child.stdin_behavior = .Ignore;
+    child.stdout_behavior = .Ignore;
+    child.stderr_behavior = .Pipe;
+
+    try child.spawn();
+
+    const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+
+    const term = child.wait() catch |err| {
+        log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+        return error.UnableToSpawnSelf;
     };
-    defer stderr_context.data.deinit();
-    var stdout_context: LLDContext = .{
-        .wasm = self,
-        .data = std.ArrayList(u8).init(self.base.allocator),
-    };
-    defer stdout_context.data.deinit();
-    const llvm = @import("../llvm.zig");
-    const ok = llvm.Link(
-        .Wasm,
-        new_argv.ptr,
-        new_argv.len,
-        append_diagnostic,
-        @ptrToInt(&stdout_context),
-        @ptrToInt(&stderr_context),
-    );
-    if (stderr_context.oom or stdout_context.oom) return error.OutOfMemory;
-    if (stdout_context.data.items.len != 0) {
-        std.log.warn("unexpected LLD stdout: {}", .{stdout_context.data.items});
+
+    switch (term) {
+        .Exited => |code| {
+            if (code != 0) {
+                // TODO parse this output and surface with the Compilation API rather than
+                // directly outputting to stderr here.
+                std.debug.print("{s}", .{stderr});
+                return error.LLDReportedFailure;
+            }
+        },
+        else => {
+            log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+            return error.LLDCrashed;
+        },
     }
-    if (!ok) {
-        // TODO parse this output and surface with the Compilation API rather than
-        // directly outputting to stderr here.
-        std.debug.print("{}", .{stderr_context.data.items});
-        return error.LLDReportedFailure;
-    }
-    if (stderr_context.data.items.len != 0) {
-        std.log.warn("unexpected LLD stderr: {}", .{stderr_context.data.items});
+
+    if (stderr.len != 0) {
+        std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
     }
 
     if (!self.base.options.disable_lld_caching) {
@@ -454,20 +459,6 @@ fn linkWithLLD(self: *Wasm, comp: *Compilation) !void {
         // other processes clobbering it.
         self.base.lock = man.toOwnedLock();
     }
-}
-
-const LLDContext = struct {
-    data: std.ArrayList(u8),
-    wasm: *Wasm,
-    oom: bool = false,
-};
-
-fn append_diagnostic(context: usize, ptr: [*]const u8, len: usize) callconv(.C) void {
-    const lld_context = @intToPtr(*LLDContext, context);
-    const msg = ptr[0..len];
-    lld_context.data.appendSlice(msg) catch |err| switch (err) {
-        error.OutOfMemory => lld_context.oom = true,
-    };
 }
 
 /// Get the current index of a given Decl in the function list

--- a/src/link/Wasm.zig
+++ b/src/link/Wasm.zig
@@ -403,46 +403,60 @@ fn linkWithLLD(self: *Wasm, comp: *Compilation) !void {
     }
 
     // Sadly, we must run LLD as a child process because it does not behave
-    // properly as a library. One exception is if we are running in passthrough
-    // mode, which means Clang / LLD should inherit stdio and are allowed to
-    // crash zig directly.
-    if (comp.clang_passthrough_mode) {
-        return @import("../main.zig").punt_to_lld(arena, argv.items);
-    }
-
+    // properly as a library.
     const child = try std.ChildProcess.init(argv.items, arena);
     defer child.deinit();
 
-    child.stdin_behavior = .Ignore;
-    child.stdout_behavior = .Ignore;
-    child.stderr_behavior = .Pipe;
+    if (comp.clang_passthrough_mode) {
+        child.stdin_behavior = .Inherit;
+        child.stdout_behavior = .Inherit;
+        child.stderr_behavior = .Inherit;
 
-    try child.spawn();
+        const term = child.spawnAndWait() catch |err| {
+            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+            return error.UnableToSpawnSelf;
+        };
+        switch (term) {
+            .Exited => |code| {
+                if (code != 0) {
+                    // TODO https://github.com/ziglang/zig/issues/6342
+                    std.process.exit(1);
+                }
+            },
+            else => std.process.abort(),
+        }
+    } else {
+        child.stdin_behavior = .Ignore;
+        child.stdout_behavior = .Ignore;
+        child.stderr_behavior = .Pipe;
 
-    const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
+        try child.spawn();
 
-    const term = child.wait() catch |err| {
-        log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
-        return error.UnableToSpawnSelf;
-    };
+        const stderr = try child.stderr.?.reader().readAllAlloc(arena, 10 * 1024 * 1024);
 
-    switch (term) {
-        .Exited => |code| {
-            if (code != 0) {
-                // TODO parse this output and surface with the Compilation API rather than
-                // directly outputting to stderr here.
-                std.debug.print("{s}", .{stderr});
-                return error.LLDReportedFailure;
-            }
-        },
-        else => {
-            log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
-            return error.LLDCrashed;
-        },
-    }
+        const term = child.wait() catch |err| {
+            log.err("unable to spawn {s}: {s}", .{ argv.items[0], @errorName(err) });
+            return error.UnableToSpawnSelf;
+        };
 
-    if (stderr.len != 0) {
-        std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+        switch (term) {
+            .Exited => |code| {
+                if (code != 0) {
+                    // TODO parse this output and surface with the Compilation API rather than
+                    // directly outputting to stderr here.
+                    std.debug.print("{s}", .{stderr});
+                    return error.LLDReportedFailure;
+                }
+            },
+            else => {
+                log.err("{s} terminated with stderr:\n{s}", .{ argv.items[0], stderr });
+                return error.LLDCrashed;
+            },
+        }
+
+        if (stderr.len != 0) {
+            std.log.warn("unexpected LLD stderr:\n{s}", .{stderr});
+        }
     }
 
     if (!self.base.options.disable_lld_caching) {

--- a/src/llvm.zig
+++ b/src/llvm.zig
@@ -1,15 +1,15 @@
 //! We do this instead of @cImport because the self-hosted compiler is easier
 //! to bootstrap if it does not depend on translate-c.
 
-pub const Link = ZigLLDLink;
-extern fn ZigLLDLink(
-    oformat: ObjectFormatType,
-    args: [*:null]const ?[*:0]const u8,
-    arg_count: usize,
-    append_diagnostic: fn (context: usize, ptr: [*]const u8, len: usize) callconv(.C) void,
-    context_stdout: usize,
-    context_stderr: usize,
-) bool;
+extern fn ZigLLDLinkCOFF(argc: c_int, argv: [*:null]const ?[*:0]const u8, can_exit_early: bool) c_int;
+extern fn ZigLLDLinkELF(argc: c_int, argv: [*:null]const ?[*:0]const u8, can_exit_early: bool) c_int;
+extern fn ZigLLDLinkMachO(argc: c_int, argv: [*:null]const ?[*:0]const u8, can_exit_early: bool) c_int;
+extern fn ZigLLDLinkWasm(argc: c_int, argv: [*:null]const ?[*:0]const u8, can_exit_early: bool) c_int;
+
+pub const LinkCOFF = ZigLLDLinkCOFF;
+pub const LinkELF = ZigLLDLinkELF;
+pub const LinkMachO = ZigLLDLinkMachO;
+pub const LinkWasm = ZigLLDLinkWasm;
 
 pub const ObjectFormatType = extern enum(c_int) {
     Unknown,

--- a/src/main.zig
+++ b/src/main.zig
@@ -176,6 +176,12 @@ pub fn mainArgs(gpa: *Allocator, arena: *Allocator, args: []const []const u8) !v
         mem.eql(u8, cmd, "-cc1") or mem.eql(u8, cmd, "-cc1as"))
     {
         return punt_to_clang(arena, args);
+    } else if (mem.eql(u8, cmd, "ld.lld") or
+        mem.eql(u8, cmd, "ld64.lld") or
+        mem.eql(u8, cmd, "lld-link") or
+        mem.eql(u8, cmd, "wasm-ld"))
+    {
+        return punt_to_lld(arena, args);
     } else if (mem.eql(u8, cmd, "build")) {
         return cmdBuild(gpa, arena, cmd_args);
     } else if (mem.eql(u8, cmd, "fmt")) {
@@ -2816,6 +2822,39 @@ fn punt_to_clang(arena: *Allocator, args: []const []const u8) error{OutOfMemory}
     }
     argv[args.len] = null;
     const exit_code = ZigClang_main(@intCast(c_int, args.len), argv[0..args.len :null].ptr);
+    process.exit(@bitCast(u8, @truncate(i8, exit_code)));
+}
+
+/// The first argument determines which backend is invoked. The options are:
+/// * `ld.lld` - ELF
+/// * `ld64.lld` - Mach-O
+/// * `lld-link` - COFF
+/// * `wasm-ld` - WebAssembly
+/// TODO https://github.com/ziglang/zig/issues/3257
+pub fn punt_to_lld(arena: *Allocator, args: []const []const u8) error{OutOfMemory} {
+    if (!build_options.have_llvm)
+        fatal("`zig {s}` unavailable: compiler built without LLVM extensions", .{args[0]});
+    // Convert the args to the format LLD expects.
+    // We subtract 1 to shave off the zig binary from args[0].
+    const argv = try arena.allocSentinel(?[*:0]const u8, args.len - 1, null);
+    for (args[1..]) |arg, i| {
+        argv[i] = try arena.dupeZ(u8, arg); // TODO If there was an argsAllocZ we could avoid this allocation.
+    }
+    const exit_code = rc: {
+        const llvm = @import("llvm.zig");
+        const argc = @intCast(c_int, argv.len);
+        if (mem.eql(u8, args[1], "ld.lld")) {
+            break :rc llvm.LinkELF(argc, argv.ptr, true);
+        } else if (mem.eql(u8, args[1], "ld64.lld")) {
+            break :rc llvm.LinkMachO(argc, argv.ptr, true);
+        } else if (mem.eql(u8, args[1], "lld-link")) {
+            break :rc llvm.LinkCOFF(argc, argv.ptr, true);
+        } else if (mem.eql(u8, args[1], "wasm-ld")) {
+            break :rc llvm.LinkWasm(argc, argv.ptr, true);
+        } else {
+            unreachable;
+        }
+    };
     process.exit(@bitCast(u8, @truncate(i8, exit_code)));
 }
 

--- a/src/zig_llvm.cpp
+++ b/src/zig_llvm.cpp
@@ -1056,39 +1056,24 @@ bool ZigLLVMWriteArchive(const char *archive_name, const char **file_names, size
     return false;
 }
 
+int ZigLLDLinkCOFF(int argc, const char **argv, bool can_exit_early) {
+    std::vector<const char *> args(argv, argv + argc);
+    return lld::coff::link(args, can_exit_early, llvm::outs(), llvm::errs());
+}
 
-bool ZigLLDLink(ZigLLVM_ObjectFormatType oformat, const char **args, size_t arg_count,
-        void (*append_diagnostic)(void *, const char *, size_t),
-        void *context_stdout, void *context_stderr)
-{
-    ArrayRef<const char *> array_ref_args(args, arg_count);
+int ZigLLDLinkELF(int argc, const char **argv, bool can_exit_early) {
+    std::vector<const char *> args(argv, argv + argc);
+    return lld::elf::link(args, can_exit_early, llvm::outs(), llvm::errs());
+}
 
-    MyOStream diag_stdout(append_diagnostic, context_stdout);
-    MyOStream diag_stderr(append_diagnostic, context_stderr);
+int ZigLLDLinkMachO(int argc, const char **argv, bool can_exit_early) {
+    std::vector<const char *> args(argv, argv + argc);
+    return lld::mach_o::link(args, can_exit_early, llvm::outs(), llvm::errs());
+}
 
-    switch (oformat) {
-        case ZigLLVM_UnknownObjectFormat:
-        case ZigLLVM_XCOFF:
-            assert(false); // unreachable
-            break;
-
-        case ZigLLVM_COFF:
-            return lld::coff::link(array_ref_args, false, diag_stdout, diag_stderr);
-
-        case ZigLLVM_ELF:
-            return lld::elf::link(array_ref_args, false, diag_stdout, diag_stderr);
-
-        case ZigLLVM_MachO:
-            return lld::mach_o::link(array_ref_args, false, diag_stdout, diag_stderr);
-
-        case ZigLLVM_Wasm:
-            return lld::wasm::link(array_ref_args, false, diag_stdout, diag_stderr);
-
-        default:
-            break;
-    }
-    assert(false); // unreachable
-    abort();
+int ZigLLDLinkWasm(int argc, const char **argv, bool can_exit_early) {
+    std::vector<const char *> args(argv, argv + argc);
+    return lld::wasm::link(args, can_exit_early, llvm::outs(), llvm::errs());
 }
 
 static AtomicRMWInst::BinOp toLLVMRMWBinOp(enum ZigLLVM_AtomicRMWBinOp BinOp) {

--- a/src/zig_llvm.h
+++ b/src/zig_llvm.h
@@ -507,9 +507,10 @@ ZIG_EXTERN_C const char *ZigLLVMGetVendorTypeName(enum ZigLLVM_VendorType vendor
 ZIG_EXTERN_C const char *ZigLLVMGetOSTypeName(enum ZigLLVM_OSType os);
 ZIG_EXTERN_C const char *ZigLLVMGetEnvironmentTypeName(enum ZigLLVM_EnvironmentType abi);
 
-ZIG_EXTERN_C bool ZigLLDLink(enum ZigLLVM_ObjectFormatType oformat, const char **args, size_t arg_count,
-        void (*append_diagnostic)(void *, const char *, size_t),
-        void *context_stdout, void *context_stderr);
+ZIG_EXTERN_C int ZigLLDLinkCOFF(int argc, const char **argv, bool can_exit_early);
+ZIG_EXTERN_C int ZigLLDLinkELF(int argc, const char **argv, bool can_exit_early);
+ZIG_EXTERN_C int ZigLLDLinkMachO(int argc, const char **argv, bool can_exit_early);
+ZIG_EXTERN_C int ZigLLDLinkWasm(int argc, const char **argv, bool can_exit_early);
 
 ZIG_EXTERN_C bool ZigLLVMWriteArchive(const char *archive_name, const char **file_names, size_t file_name_count,
         enum ZigLLVM_OSType os_type);

--- a/test/cli.zig
+++ b/test/cli.zig
@@ -92,13 +92,13 @@ fn exec(cwd: []const u8, expect_0: bool, argv: []const []const u8) !ChildProcess
 fn testZigInitLib(zig_exe: []const u8, dir_path: []const u8) !void {
     _ = try exec(dir_path, true, &[_][]const u8{ zig_exe, "init-lib" });
     const test_result = try exec(dir_path, true, &[_][]const u8{ zig_exe, "build", "test" });
-    testing.expect(std.mem.endsWith(u8, test_result.stderr, "All 1 tests passed.\n"));
+    testing.expectStringEndsWith(test_result.stderr, "All 1 tests passed.\n");
 }
 
 fn testZigInitExe(zig_exe: []const u8, dir_path: []const u8) !void {
     _ = try exec(dir_path, true, &[_][]const u8{ zig_exe, "init-exe" });
     const run_result = try exec(dir_path, true, &[_][]const u8{ zig_exe, "build", "run" });
-    testing.expect(std.mem.eql(u8, run_result.stderr, "info: All your codebase are belong to us.\n"));
+    testing.expectEqualStrings("info: All your codebase are belong to us.\n", run_result.stderr);
 }
 
 fn testGodboltApi(zig_exe: []const u8, dir_path: []const u8) anyerror!void {

